### DISCUSSION TITLE
Automate user registration

### DISCUSSION THREAD

**Discussion by prathameshbobhate** [zitadel/zitadel#7111] [General]
Hi All,
I am new to Zitadel. We are testing Netbird for our organization. We are using OIDC for SSO authentication. 
When the user login using IDP it says "external user not found. Do you want to link your user or auto-register a new one?"  We want to auto-register the user with the information provided by IDP. Could anyone guide me on this?

**Comment by hifabienne** [comment]
When you register an external identity provider, you can enable automatic creation:
<img width="1196" alt="image" src="https://github.com/zitadel/zitadel/assets/38692350/1648af49-f049-42d9-ac20-929543b3b533">

To be able to register users automatically, it requires all mandatory fields. The fields are taken from the token we get from the provider. If a mandatory field is not filled, you can fill it yourself with a ZITADEL action.

**Reply by gabriel-hoff-vectorsolutions** [reply]
@hifabienne We are running into a similar issue when using a SAML Identity Provider:

![image](https://github.com/zitadel/zitadel/assets/62257776/d09bc550-50f6-4a39-815e-9150ef605d30)

Questions:
- What fields would have to be present in the SAML assertion for the user to be automatically created?
- If an action is needed to fill in the missing information, do you have an example of what it should look like?

**Reply by gabriel-hoff-vectorsolutions** [reply]
If it helps, I'm using [mocksaml.com](https://mocksaml.com/) as the SAML identity provider with the following optional settings:
![image](https://github.com/zitadel/zitadel/assets/62257776/095610da-09e5-4f2e-b5b1-34bed9d85ed6)

**Reply by gabriel-hoff-vectorsolutions** [reply]
I've tried creating an action following [this example](https://github.com/zitadel/actions/blob/main/examples/post_auth_saml.js), but I'm still getting the same error message. Is automatic user creation supported with SAML IDPs?

**Reply by gabriel-hoff-vectorsolutions** [reply]
I was able to automatically create the user by also enabling the "Account creation allowed" option. To summarize, to make this work you have to:
1. Have a rule similar to [this one](https://github.com/zitadel/actions/blob/main/examples/post_auth_saml.js)
2. Enable "Automatic creation"
3. Enable "Account creation allowed"

**Reply by hifabienne** [reply]
Hei @vs-gtadeu 
Yes sorry didn't mentioned that explicitly. 
The creation must be allowed, other wise auto creation doesn't work.
So if I understand you right it works now?

**Reply by gabriel-hoff-vectorsolutions** [reply]
Yes, it is working right now. Thank you!

**Reply by cmcrystle** [reply]
Is there a script to automate linking users to an existing account?  Creating the users accounts and linking to it when they log into via SAML?

**Reply by hifabienne** [reply]
in the idp you have a setting auto linking, if all requried fields are filled auto linking should work.
in zitadel firstname, lastname and email are required. if those fields are not automatically filled you can use an action to prefill the data to make autolinking possible

**Comment by cmcrystle** [comment]
[2025-02-04_11-50-17.pdf](https://github.com/user-attachments/files/18660567/2025-02-04_11-50-17.pdf)

Are you referring to the disabling the manual account linking and then using the email or user name for the linked account?

we are trying to create the users and add them to the linked application before they access the system and it doesn't seem to work correctly.

**Reply by cmcrystle** [reply]
Do I need to uncheck the auto create option, so that it defaults to linking?

**Reply by hifabienne** [reply]
No you don't need to unchekc the auto creation option. If there is a user who can be linked, it should be linked with the auto linking option. But if no user can be found it would be automatically created if you have enabeled that option.

From what I read you are talking now about different things, one is linking a user, which is about authentication only, and can be done with those options.
The second thing is add them to the linked application, what do you mean by that? Give the user a role for a app/project? In that case you would need an Action to automatically create those

**Reply by cmcrystle** [reply]
Yes -- I was able to have the user information updated and the identity provider assigned to the user record.  We were trying to avoid the user interaction with zitadel and direct them to the application using zitadel.  Do you have any ideas on how to turn off the password requirement?

**Reply by hifabienne** [reply]
What do you mean by password requirements? Password length and characters? There is a password complexity setting you can configure

**Comment by cmcrystle** [comment]
I want to bypass the password requirement all together.  We have it set up for SSO, but the user must create a new password the first time they access the application.

**Reply by hifabienne** [reply]
hm, generally if you create a user trhough sso, password should not be requested. But if you have already created user before maybe with an initial password, which is set to be updated on the next login?
